staging/src/k8s.io/client-go/tools/cache/reflector_watchlist_test.go

```diff
@@ -0,0 +1,441 @@
+		},
+		{
+			name:                  "check if Updates and Deletes events are propagated during streaming (until the bookmark is received)",
+			closeAfterWatchEvents: 6,
+			returnedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: makePod("p1", "1")},
+				{Type: watch.Added, Object: makePod("p2", "2")},
+				{Type: watch.Modified, Object: func() runtime.Object {
+					p1 := makePod("p1", "3")
+					p1.Spec.ActiveDeadlineSeconds = pointer.Int64(12)
+					return p1
+				}()},
+				{Type: watch.Added, Object: makePod("p3", "4")},
+				{Type: watch.Deleted, Object: makePod("p3", "5")},
+				{Type: watch.Bookmark, Object: &v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						ResourceVersion: "5",
+						Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+					},
+				}},
+			},
+			expectedWatchRequests: 1,
+			expectedRequestOptions: []metav1.ListOptions{
+				{SendInitialEvents: pointer.Bool(true), AllowWatchBookmarks: true, ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan, TimeoutSeconds: pointer.Int64(1)},
+			},
+			expectedStoreContent: []v1.Pod{
+				*makePod("p2", "2"),
+				func() v1.Pod {
+					p1 := *makePod("p1", "3")
+					p1.Spec.ActiveDeadlineSeconds = pointer.Int64(12)
+					return p1
+				}(),
+			},
+		},
+		{
+			name: "checks if the reflector retries 429",
+			watchServerErrorPredicate: func() func(options metav1.ListOptions) error {
+				counter := 1
+				return func(options metav1.ListOptions) error {
+					if counter < 3 {
+						counter++
+						return apierrors.NewTooManyRequests("busy, check again later", 1)
+					}
+					return nil
+				}
+			}(),
+			closeAfterWatchEvents: 2,
+			returnedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: makePod("p1", "1")},
+				{Type: watch.Bookmark, Object: &v1.Pod{
+					ObjectMeta: metav1.ObjectMeta{
+						ResourceVersion: "2",
+						Annotations:     map[string]string{"k8s.io/initial-events-end": "true"},
+					},
+				}},
+			},
+			expectedWatchRequests: 3,
+			expectedRequestOptions: []metav1.ListOptions{
+				{SendInitialEvents: pointer.Bool(true), AllowWatchBookmarks: true, ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan, TimeoutSeconds: pointer.Int64(1)},
+				{SendInitialEvents: pointer.Bool(true), AllowWatchBookmarks: true, ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan, TimeoutSeconds: pointer.Int64(1)},
+				{SendInitialEvents: pointer.Bool(true), AllowWatchBookmarks: true, ResourceVersionMatch: metav1.ResourceVersionMatchNotOlderThan, TimeoutSeconds: pointer.Int64(1)},
+			},
+			expectedStoreContent: []v1.Pod{*makePod("p1", "1")},
+		},
+		{
+			name:                  "check if stopping a watcher before sync results in creating a new watch-list request",
+			stopAfterWatchEvents:  1,
+			closeAfterWatchEvents: 3,
+			returnedWatchEvents: []watch.Event{
+				{Type: watch.Added, Object: makePod("p1", "1")},
+				// second request
+				{Type: watch.Added, Object: makePod("p2", "2")},
```

Comment:
I would sent the above p1 event again here - so that it looks more like a retry.